receiver/oracledbreceiver/scraper.go

```diff
@@ -0,0 +1,395 @@
+	pgaMemory               = "session pga memory"
+	sessionCountSQL         = "select status, type, count(*) as VALUE FROM v$session GROUP BY status, type"
+	systemResourceLimitsSQL = "select RESOURCE_NAME, CURRENT_UTILIZATION, LIMIT_VALUE, CASE WHEN TRIM(INITIAL_ALLOCATION) LIKE 'UNLIMITED' THEN '-1' ELSE TRIM(INITIAL_ALLOCATION) END as INITIAL_ALLOCATION, CASE WHEN TRIM(LIMIT_VALUE) LIKE 'UNLIMITED' THEN '-1' ELSE TRIM(LIMIT_VALUE) END as LIMIT_VALUE from v$resource_limit"
+	tablespaceUsageSQL      = "select TABLESPACE_NAME, BYTES from DBA_DATA_FILES"
+	tablespaceMaxSpaceSQL   = "select TABLESPACE_NAME, (BLOCK_SIZE*MAX_EXTENTS) AS VALUE FROM DBA_TABLESPACES"
+)
+
+type dbProviderFunc func() (*sql.DB, error)
+
+type clientProviderFunc func(*sql.DB, string, *zap.Logger) dbClient
+
+type scraper struct {
+	statsClient                dbClient
+	tablespaceMaxSpaceClient   dbClient
+	tablespaceUsageClient      dbClient
+	systemResourceLimitsClient dbClient
+	sessionCountClient         dbClient
+	db                         *sql.DB
+	clientProviderFunc         clientProviderFunc
+	metricsBuilder             *metadata.MetricsBuilder
+	dbProviderFunc             dbProviderFunc
+	logger                     *zap.Logger
+	id                         config.ComponentID
+	instanceName               string
+	scrapeCfg                  scraperhelper.ScraperControllerSettings
+	startTime                  pcommon.Timestamp
+	metricsSettings            metadata.MetricsSettings
+}
+
+func newScraper(id config.ComponentID, metricsBuilder *metadata.MetricsBuilder, metricsSettings metadata.MetricsSettings, scrapeCfg scraperhelper.ScraperControllerSettings, logger *zap.Logger, providerFunc dbProviderFunc, clientProviderFunc clientProviderFunc, instanceName string) *scraper {
+	return &scraper{
+		id:                 id,
+		metricsBuilder:     metricsBuilder,
+		metricsSettings:    metricsSettings,
+		scrapeCfg:          scrapeCfg,
+		logger:             logger,
+		dbProviderFunc:     providerFunc,
+		clientProviderFunc: clientProviderFunc,
+		instanceName:       instanceName,
+	}
+}
+
+var _ scraperhelper.Scraper = (*scraper)(nil)
+
+func (s *scraper) ID() config.ComponentID {
+	return s.id
+}
+
+func (s *scraper) Start(context.Context, component.Host) error {
+	s.startTime = pcommon.NewTimestampFromTime(time.Now())
+	var err error
+	s.db, err = s.dbProviderFunc()
+	if err != nil {
+		return fmt.Errorf("failed to open db connection: %w", err)
+	}
+	s.statsClient = s.clientProviderFunc(s.db, statsSQL, s.logger)
+	s.sessionCountClient = s.clientProviderFunc(s.db, sessionCountSQL, s.logger)
+	s.systemResourceLimitsClient = s.clientProviderFunc(s.db, systemResourceLimitsSQL, s.logger)
+	s.tablespaceUsageClient = s.clientProviderFunc(s.db, tablespaceUsageSQL, s.logger)
+	s.tablespaceMaxSpaceClient = s.clientProviderFunc(s.db, tablespaceMaxSpaceSQL, s.logger)
+	return nil
+}
+
+func (s *scraper) Scrape(ctx context.Context) (pmetric.Metrics, error) {
```

Comment:
There's kind of a lot of orthogonal functionality going on in this scraper. What do you think about splitting it up, perhaps into multiple, relatively small scrapers that get registered at startup time based on metric settings? That way, we'd end up with a few small types/functions and we won't have to checking setting on every scrape.

Reply:
I think having a single scraper is probably preferrable to send all metrics in one scrape and use at most one DB connection. The code could be nicer to look at, and refactored into smaller functions and appended as an array of calls that are set at start time if the setting is enabled.